network/src/commonMain/kotlin/com/wire/kalium/network/api/user/login/LoginApiImpl.kt

```diff
@@ -1,66 +1,55 @@
 import io.ktor.client.request.setBody
 import io.ktor.http.URLProtocol
 import io.ktor.http.set
-import kotlinx.serialization.SerialName
-import kotlinx.serialization.Serializable
 
 class LoginApiImpl(private val httpClient: HttpClient) : LoginApi {
 
-    @Serializable
-    internal data class LoginRequest(
-        @SerialName("email") val email: String? = null,
-        @SerialName("handle") val handle: String? = null,
-        @SerialName("password") val password: String,
-        @SerialName("label") val label: String
-    )
-
-    private fun LoginApi.LoginParam.toRequestBody(): LoginRequest {
-        return when (this) {
-            is LoginApi.LoginParam.LoginWithEmail -> LoginRequest(email = email, password = password, label = label)
-            is LoginApi.LoginParam.LoginWithHandel -> LoginRequest(handle = handle, password = password, label = label)
-        }
-    }
-
     override suspend fun login(
         param: LoginApi.LoginParam, persist: Boolean, apiBaseUrl: String
-    ): NetworkResponse<SessionDTO> {
-
-        val result = wrapKaliumResponse<AccessTokenDTO> {
-            httpClient.post {
-                url.set(host = apiBaseUrl, path = PATH_LOGIN)
-                url.protocol = URLProtocol.HTTPS
+    ): NetworkResponse<AuthenticationResult> = wrapKaliumResponse<AccessTokenDTO> {
+        httpClient.post {
+            url.set(host = apiBaseUrl, path = PATH_LOGIN)
+            url.protocol = URLProtocol.HTTPS
 
-                parameter(QUERY_PERSIST, persist)
-                setBody(param.toRequestBody())
-            }
+            parameter(QUERY_PERSIST, persist)
+            setBody(param.toBody())
         }
-        return when (result) {
-            is NetworkResponse.Success -> {
-                val refreshToken = result.cookies[RefreshTokenProperties.COOKIE_NAME]
-                if (refreshToken == null) {
-                    CustomErrors.MISSING_REFRESH_TOKEN
-                } else {
-                    NetworkResponse.Success(result.value.toSessionDto(refreshToken), result.headers, result.httpCode)
-                }
-            }
-            is NetworkResponse.Error -> NetworkResponse.Error(result.kException)
+    }.flatMap { accessTokenDTOResponse ->
+        with(accessTokenDTOResponse) {
+            cookies[RefreshTokenProperties.COOKIE_NAME]?.let { refreshToken ->
+                NetworkResponse.Success(value.toSessionDto(refreshToken), headers, httpCode)
+            } ?: CustomErrors.MISSING_REFRESH_TOKEN
         }
+    }.flatMap { sessionDTOResponse ->
+        wrapKaliumResponse<UserDTO> {
+            httpClient.get {
+                url.set(host = apiBaseUrl, path = PATH_SELF)
+                url.protocol = URLProtocol.HTTPS
+                bearerAuth(sessionDTOResponse.value.accessToken)
+            }
+        }.mapSuccess { userDTO -> AuthenticationResult(sessionDTOResponse.value, userDTO) }
     }
 
 
     private companion object {
+        const val PATH_SELF = "self"
```

Comment:
Can't you use the existing `SelfApi` instead of extending `LoginApi`?

Reply:
In the `SelfApi` it uses an `AuthnticatedHttpClient` which add the BE url and `Authorization` header from local storage, and at this point there is no session stored to the User, so both have to be done manually 
```kotlin 
            httpClient.get {
                url.set(host = apiBaseUrl, path = PATH_SELF)
                url.protocol = URLProtocol.HTTPS
                bearerAuth(sessionDTOResponse.value.accessToken)
            }
```
and I assumed that this hack will be removed once BE return a QualifiedID after Login

Reply:
> I assumed that this hack will be removed once BE return a QualifiedID after Login

Right, then this makes sense.